### PR TITLE
add logger in explainer and transformer

### LINKED ISSUE #1404: Logger container not being created for explainer spec
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Deployed the [AIX explainer](https://github.com/kubeflow/kfserving/tree/master/docs/samples/explanation/aix/mnist#using-aix-to-get-explanations-for-mnist-classifications) and added a logger 
```
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "InferenceService"
metadata:
  name: "aix-explainer"
  namespace: default
spec:
  predictor:
    containers:
    - name: predictor
      image: aipipeline/rf-predictor:0.4.1
      command: ["python", "-m", "rfserver", "--model_name", "aix-explainer"]
      imagePullPolicy: Always
  explainer:
    logger:
      url: http://message-dumper.default/
      mode: all
    aix:
      type: LimeImages
      config:
        num_samples: "100"
        top_labels: "10"
        min_weight: "0.01"

```

But when I check on the pods I notice the `aix-explainer-explainer-default-dv98t-deployment-68d8755f6dlrr9` has `3/3` with no logger deployed
```
bash-3.2$ kubectl get pods -w
NAME                                                              READY   STATUS    RESTARTS   AGE
aix-explainer-explainer-default-dv98t-deployment-68d8755f6dlrr9   3/3     Running   0          3m13s
aix-explainer-predictor-default-rn6rw-deployment-75bcf8c8fn9ch9   3/3     Running   0          3m12s
message-dumper-j85sv-deployment-6db5fcfd47-mq8jq                  3/3     Running   0          101m
```
And after querying the aix-explainer, no logs can are in the message dumper
```
bash-3.2$ kubectl logs $(kubectl get pod -l serving.knative.dev/service=message-dumper -o jsonpath='{.items[0].metadata.name}') user-container
bash-3.2$ 
```

**What did you expect to happen:**

Similar to what we see when the logger is deployed to the predictor
```
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "InferenceService"
metadata:
  name: "aix-explainer"
  namespace: default
spec:
  predictor:
    logger:
      url: http://message-dumper.default/
      mode: all
    containers:
    - name: predictor
      image: aipipeline/rf-predictor:0.4.1
      command: ["python", "-m", "rfserver", "--model_name", "aix-explainer"]
      imagePullPolicy: Always
  explainer:
    aix:
      type: LimeImages
      config:
        num_samples: "100"
        top_labels: "10"
        min_weight: "0.01"
```

```
bash-3.2$ kubectl get pods -w
NAME                                                              READY   STATUS    RESTARTS   AGE
aix-explainer-explainer-default-c9wfk-deployment-8cbbcbdb7qc9d2   3/3     Running   0          4m50s
aix-explainer-predictor-default-psd8g-deployment-db8f584fdzhxgj   4/4     Running   0          4m49s
message-dumper-j85sv-deployment-6db5fcfd47-mq8jq                  3/3     Running   0          123m
```
And checking the message dumper after querying the predictor results in receiving the cloudevents created from the requests
```
bash-3.2$ kubectl logs $(kubectl get pod -l serving.knative.dev/service=message-dumper -o jsonpath='{.items[0].metadata.name}') user-container
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: org.kubeflow.serving.inference.request
  source: http://localhost:9081/
  id: 2062b33a-74ed-4b36-b512-9975ff4fe99f
  time: 2021-02-23T23:44:18.143788921Z
  datacontenttype: application/json
Extensions,
  endpoint: 
  inferenceservicenam
...
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version:
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.17+IKS", GitCommit:"8f82fcbe37a8086624922025f586f48b302f0f28", GitTreeState:"clean", BuildDate:"2021-01-15T02:17:32Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from `/etc/os-release`): `macOS Big Sur 11.2`

### PR DESCRIPTION
**Pull request by theofpa**
**What this PR does / why we need it**:
Adds the logging functionality in explainer, by adding annotations for logger in the reconciler of the explainer, similarly to the predictor logger.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1404

**Special notes for your reviewer**:
Test results:
```
$ curl -H 'Host: german-credit.default.example.com' -H 'Content-type: application/json' -d@./input.json http://172.18.255.200/v1/models/german-credit:explain 
{"predictions": [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], "metrics": {"base_rate": 0.9230769230769231, "consistency": [0.9871794871794872], "disparate_impact": 0.45454545454545453, "num_instances": 78.0, "num_negatives": 6.0, "num_positives": 72.0, "statistical_parity_difference": -0.5454545454545454}}
```

```
$ kubectl logs message-dumper-00001-deployment-5764b464c5-trr9f --tail 12 -c user-container
    "metrics": {
      "base_rate": 0.9230769230769231,
      "consistency": [
        0.9871794871794872
      ],
      "disparate_impact": 0.45454545454545453,
      "num_instances": 78.0,
      "num_negatives": 6.0,
      "num_positives": 72.0,
      "statistical_parity_difference": -0.5454545454545454
    }
  }
```

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Logging capability on explainers
```

### CODE REVIEW

**Comment by yuzisun** [comment]
@theofpa How about transformer as well as we discussed in the summit ?

**Comment by theofpa** [comment]
> @theofpa How about transformer as well as we discussed in the summit ?

thanks Dan, I've added annotations in the transformer, as well as moved the helper functions for annotations to a better location.

Tested the logging:
```
$ curl -H 'Content-Type: application/json' -H 'Host: torchserve-transformer.default.example.com' -X POST -d@docs/samples/v1beta1/transformer/torchserve_image_transformer/input.json http://172.18.255.200/v1/models/mnist:predict
{"predictions": [2]}
```
```
$ kubectl logs message-dumper-00001-deployment-5764b464c5-p9ptf -c user-container | head -23
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: org.kubeflow.serving.inference.request
  source: http://localhost:9081/
  id: cefcffda-f404-44d9-b7b5-a3acffc846f0
  time: 2021-05-06T12:31:04.311596983Z
  datacontenttype: application/json
Extensions,
  endpoint: 
  inferenceservicename: torchserve-transformer
  namespace: default
  traceparent: 00-4bbebfd72514f1cec3d37f2704ea1359-7264cf4d02b15c41-00
Data,
  {
    "instances": [
      {
        "data": "iVBORw0KGgoAAAANSUhEUgAAABwAAAAcCAAAAABXZoBIAAAAw0lEQVR4nGNgGFggVVj4/y8Q2GOR83n+58/fP0DwcSqmpNN7oOTJw6f+/H2pjUU2JCSEk0EWqN0cl828e/FIxvz9/9cCh1zS5z9/G9mwyzl/+PNnKQ45nyNAr9ThMHQ/UG4tDofuB4bQIhz6fIBenMWJQ+7Vn7+zeLCbKXv6z59NOPQVgsIcW4QA9YFi6wNQLrKwsBebW/68DJ388Nun5XFocrqvIFH59+XhBAxThTfeB0r+vP/QHbuDCgr2JmOXoSsAAKK7bU3vISS4AAAAAElFTkSuQmCC",
        "target": 0
      }
    ]
  }
```

Maybe we should introduce new types in the context for the components, so now we have `type: org.kubeflow.serving.inference.request/response`, but maybe we can break it down to the following, to help on the model monitoring from payload logging:
* `type: org.kubeflow.serving.predict.request/response`
* `type: org.kubeflow.serving.transform.request/response`
* `type: org.kubeflow.serving.explain.request/response`

What do you think, @yuzisun @drewbutlerbb4 ?

**Comment by drewbutlerbb4** [comment]
Thanks @theofpa this looks awesome!
I could definitely see the added types being useful in the case where you have both an explainer and predictor deployed with logging enabled for both.

File: pkg/controller/v1beta1/inferenceservice/components/component.go [hunk 1]
```diff
@@ -13,9 +13,90 @@ limitations under the License.
...
+		if logger.URL != nil {
+			annotations[constants.LoggerSinkUrlInternalAnnotationKey] = *logger.URL
+		}
+		annotations[constants.LoggerModeInternalAnnotationKey] = string(logger.Mode)
+		return true
+	}
+	return false
+}
+
+func addLoggerContainerPort(container *v1.Container) {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/component.go:42]: `addLoggerContainerPort` and `addBatcherContainerPort` is the same function now as they are all merged to agent container, can you merge these to `addAgentContainerPort`?

File: pkg/controller/v1beta1/inferenceservice/components/transformer.go [hunk 2]
```diff
@@ -62,6 +62,8 @@ func (p *Transformer) Reconcile(isvc *v1beta1.InferenceService) error {
 	if sourceURI := transformer.GetStorageUri(); sourceURI != nil {
 		annotations[constants.StorageInitializerSourceUriInternalAnnotationKey] = *sourceURI
 	}
+	hasInferenceLogging := addLoggerAnnotations(isvc.Spec.Transformer.Logger, annotations)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/transformer.go:65]:
```go
hasLogger := addLoggerAnnotations(isvc.Spec.Transformer.Logger, annotations)
hasBatcher := addLoggerAnnotations(isvc.Spec.Transformer.Batcher, annotations)
```

File: pkg/controller/v1beta1/inferenceservice/components/transformer.go [hunk 3]
```diff
@@ -83,6 +85,9 @@ func (p *Transformer) Reconcile(isvc *v1beta1.InferenceService) error {
 		isvc.Spec.Transformer.PodSpec.Containers[0] = *container
 	}
 
+	if hasInferenceLogging {
+		addLoggerContainerPort(&isvc.Spec.Transformer.PodSpec.Containers[0])
+	}
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/components/transformer.go:90]:
```go
if hasLogger || hasBatcher {
    addAgentContainerPort(&isvc.Spec.Transformer.PodSpec.Containers[0])
}
```

File: pkg/controller/v1beta1/inferenceservice/components/component.go (hunk 1 above)
Comment (theofpa) [reply pkg/controller/v1beta1/inferenceservice/components/component.go]: 0cfefe9

File: pkg/controller/v1beta1/inferenceservice/components/transformer.go (hunk 3 above)
Comment (theofpa) [reply pkg/controller/v1beta1/inferenceservice/components/transformer.go:91]: 0cfefe9

File: pkg/controller/v1beta1/inferenceservice/components/transformer.go (hunk 2 above)
Comment (theofpa) [reply pkg/controller/v1beta1/inferenceservice/components/transformer.go:65]: 0cfefe9

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve

**Pull request merged**